src/main/java/house/inksoftware/systemtest/domain/steps/response/rest/ExpectedRestResponseStep.java

```diff
@@ -1,29 +1,251 @@
+    }
+
+    @NotNull
+    private HashSet<String> findAllFieldsToBeComparedNormally() {
+        var result = new HashSet<>(allFieldsSet);
+        result.removeAll(verificationFieldsSet);
+
+        return result;
+    }
+
+    private void compareUsingVerification(JsonNode actualResponseBody) {
+        var verificationSteps = verification.elements();
+
+        verificationSteps.forEachRemaining(verificationStep -> {
+
+            var field1 = getNodeByAttributePath(body, verificationStep.path("field").asText());
+            var field2 = getNodeByAttributePath(actualResponseBody, verificationStep.path("field").asText());
+
+            var verificationType = verificationStep.path("type").asText();
+            if (verificationType.equals("cosine-similarity")) {
+                compareByCosineSimilarity(verificationStep, field1, field2);
+            } else {
+                assertEquals(field1, field2);
+            }
+        });
+    }
+
+    private static Set<String> findAllFieldNamesFromJSON(Set<String> result, String currentField, JsonNode node) {
+        var isLeaf = !node.isContainerNode();
+
+        if (isLeaf) {
+            result.add(currentField);
+            return result;
+        }
+        return pathFurtherNodes(result, currentField, node);
+    }
+
+    private static Set<String> pathFurtherNodes(Set<String> result, String currentField, JsonNode node) {
+        var isArrayNode = node.isArray();
+
+        if (isArrayNode) {
+            return findAllFieldsFromJSONArray(result, currentField, node);
+        }
+        return findAllFieldsFromJSONDictionary(result, currentField, node);
+    }
+
+    private static Set<String> findAllFieldsFromJSONDictionary(Set<String> result, String currentField, JsonNode node) {
+        var fields = node.fieldNames();
+
+        fields.forEachRemaining(field -> {
+            var nextField = currentField.isBlank() ? field : currentField + "." + field;
+            result.addAll(findAllFieldNamesFromJSON(result, nextField, node.path(field)));
+        });
+
+        return result;
+    }
+
+    private static Set<String> findAllFieldsFromJSONArray(Set<String> result, String currentField, JsonNode node) {
+
+        IntStream.range(0, node.size()).forEach(i -> {
+            result.addAll(findAllFieldNamesFromJSON(result, currentField+"["+i+"]", node.get(i)));
+        });
+
+        return result;
+    }
+
+    private static void compareByCosineSimilarity(JsonNode verificationStep, JsonNode field1, JsonNode field2) {
+        double minSimilarityThreshold = verificationStep.path("minSimilarityThreshold").asDouble();
+
+        var vector1 = callAdaModel(field1.asText());
+        var vector2 = callAdaModel(field2.asText());
+
+        double similarity = calculateCosineSimilarity(vector1, vector2);
+
+        assertTrue("texts are not similar ("+field1.asText()+", "+field2.asText()+"), similarity: "+ similarity, similarity >= minSimilarityThreshold);
```

Comment:
formatting, lack of spaces to read clearly, also it's a sentence so please start with capital letter